shared/fs/footer/upload.native.js

```diff
@@ -1,66 +1,145 @@
 // @flow
 import * as React from 'react'
-import {globalStyles, globalColors} from '../../styles'
-import {Text, Box} from '../../common-adapters'
+import {globalStyles, globalColors, platformStyles} from '../../styles'
+import {Button, Text, Box} from '../../common-adapters'
 import {type UploadProps} from './upload'
 import {NativeAnimated, NativeEasing} from '../../common-adapters/native-wrappers.native'
 
-const patternRequire = require('../../images/upload-pattern-80.png')
+const patternRequire = require('../../images/upload-pattern-2-600.png')
 
 type UploadState = {
   backgroundTop: NativeAnimated.AnimatedValue,
+  uploadTop: NativeAnimated.AnimatedValue,
+  mounted: boolean,
 }
 
+const easing = NativeEasing.bezier(0.13, 0.72, 0.31, 0.95)
+
 class Upload extends React.PureComponent<UploadProps, UploadState> {
   state = {
     backgroundTop: new NativeAnimated.Value(0),
+    uploadTop: new NativeAnimated.Value(48),
+    mounted: false,
+  }
+
+  animations = {
+    loop: null,
+    in: null,
+    out: null,
   }
 
-  componentDidMount() {
-    NativeAnimated.loop(
+  startAnimationLoop() {
+    this.animations.loop = NativeAnimated.loop(
       NativeAnimated.timing(this.state.backgroundTop, {
-        toValue: -40,
+        toValue: -80, // pattern loops on multiples of 80
         duration: 2000,
         easing: NativeEasing.linear,
       })
-    ).start()
+    )
+    this.animations.loop.start()
+  }
+  startAnimationIn() {
+    this.animations.in = NativeAnimated.timing(this.state.uploadTop, {
+      toValue: 0,
+      duration: 300,
+      easing,
+    })
+    this.animations.in.start()
+  }
+  startAnimationOut(cbIfFinish: () => void) {
+    this.animations.out = NativeAnimated.timing(this.state.uploadTop, {
+      toValue: 48,
+      duration: 300,
+      easing,
+    })
+    this.animations.out.start(({finished}) => finished && cbIfFinish())
+  }
+  stopAnimation(animation: string) {
+    if (!this.animations[animation]) {
+      return
+    }
+    this.animations[animation].stop()
+    this.animations[animation] = null
+  }
+
+  enter() {
+    this.setState({mounted: true})
+    this.stopAnimation('out')
+    this.startAnimationIn()
+    this.startAnimationLoop()
+  }
+
+  exit() {
+    this.stopAnimation('in')
+    this.startAnimationOut(() => {
+      this.stopAnimation('loop')
+      this.setState({mounted: false})
```

Comment:
i don't believe this is safe

Reply:
Could you explain why this is not safe? It's in the callback that's called when the animation has finished cleanly.

Reply:
any asynchronous callback can't safely call setState as the component could be unmounted. you either need to bookkeep the mounting or cancel any calls like this while unmounting

Reply:
I see. What's the consequences of calling it in async and race happens?

In this case, I need this to determine whether to render the banner or not. If the whole component is unmounted, it doesn't really matter. Also, the component is always mounted as long as user is in the Files tab.

By bookkeeping, do you mean I should have a class variable that's set to `true` in `componentWillUnmount`, and only call this `setState` if that variable is `false`?

Reply:
react will give you warnings telling you its unsafe to do this. this is just a general rule. so if you do
```
setTimeout(() => this.setState({foo: 'bar'}), 10000)
```
that's unsafe.

yes, by bookkeeping i mean either

```
this.timeoutID = setTimeout(() => this.setState({foo: 'bar'}), 10000)

componentWillUnmount() {
clearTimeout(this.timeoutID)
}

// or

_mounted = true
componentWillUnmount () {
this._mounted = false
}

componentDidMount () {
this._mounted = true
}

setTimeout(() => this._mounted && this.setState({foo: 'bar'}), 10000)
```